examples/darwin-framework-tool/commands/configuration/PrintDeviceCommand.h

```diff
@@ -0,0 +1,459 @@
+/*
+ *   Copyright (c) 2025 Project CHIP Authors
+ *   All rights reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+#pragma once
+
+#import <Matter/Matter.h>
+
+#import "../common/DeviceDelegate.h"
+#import "MTRError_Utils.h"
+
+#include <commands/common/Command.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+class PrintDeviceCommand : public CHIPCommandBridge {
+public:
+    PrintDeviceCommand()
+        : CHIPCommandBridge("print", "Prints the endpoints, clusters, commands and attributes that are "
+                                     "available on a device instance.")
+    {
+        AddArgument("node-id", 0, UINT64_MAX, &mNodeId, "The Node ID of the device whose details will be printed.");
+        AddArgument("endpoint-id", 0, UINT16_MAX, &mEndPointId, "If provided, restrict the output to this single endpoint.");
+    }
+
+protected:
+    /////////// CHIPCommandBridge Interface /////////
+    CHIP_ERROR RunCommand() override
+    {
+        auto * controller = CurrentCommissioner();
+        VerifyOrReturnError(nil != controller, CHIP_ERROR_INCORRECT_STATE);
+
+        auto * device = [MTRDevice deviceWithNodeID:@(mNodeId) controller:controller];
+        VerifyOrReturnError(nil != device, CHIP_ERROR_INCORRECT_STATE);
+
+        DeviceWithNodeId(mNodeId);
+
+        if (device.deviceCachePrimed) {
+            ReturnErrorOnFailure(PrintDevice(device));
+        } else {
+            auto onDeviceCachePrimedBlock = ^{
+                CHIP_ERROR error = PrintDevice(device);
+
+                [device removeDelegate:mDeviceDelegate];
+                mDeviceDelegate = nil;
+                SetCommandExitStatus(error);
+            };
+
+            mDeviceDelegate = [[DeviceDelegate alloc] init];
+            mDeviceDelegate.onDeviceCachePrimed = onDeviceCachePrimedBlock;
+            auto deviceDelegateDispatchQueue = dispatch_queue_create("com.chip.devicedelegate", DISPATCH_QUEUE_SERIAL_WITH_AUTORELEASE_POOL);
+            [device addDelegate:mDeviceDelegate queue:deviceDelegateDispatchQueue];
+        }
+
+        SetCommandExitStatus(CHIP_NO_ERROR);
```

Comment:
Doesn't this call SetCommandExitStatus immediately and then again when the async block runs?

As in, this line should not be here; it should be in the case after we synchronously called PrintDevice only.

Reply:
Sorry...